slapo/model_schedule/gpt2.py

```diff
@@ -1,19 +1,133 @@
+    except Exception:
+        model_name = model_config.get("_name_or_path", None)
+    logger = get_logger(f"{model_name}")
+
+    # Change data type.
+    fp16 = sch_config.get("fp16", False)
+    bf16 = sch_config.get("bf16", False)
+    if fp16 and bf16:
+        raise ValueError("Cannot use both fp16 and bf16")
+    if fp16:
+        logger.info("Change model dtype to fp16", ranks=0)
+        model.half()
+    elif bf16:
+        logger.info("Change model dtype to bf16", ranks=0)
+        model.bfloat16()
+    else:
+        logger.info("Use fp32 as default model dtype", ranks=0)
+
+    group = sch_config.get("group", None)
+    sch = create_schedule(model, group=group)
+    logger.info(
+        "Scheduling %s with TP=%d, config: %s",
+        model_name,
+        sch.world_size,
+        sch_config,
+        ranks=0,
+    )
+
+    # Tensor parallelism.
+    logger.info("Shard model parameters", ranks=0)
+    shard_parameters(sch, model_config, sch_config)
+
+    if sch.world_size > 1 and sch_config.get("bcast_input", False):
+        # Broadcast input to all devices within the MP group.
+        # This is not required when running on Megatron.
+        logger.info("Broadcast input to all devices", ranks=0)
+        broadcast_input(sch)
+
+    # Insert activation checkpoints.
+    ckpt_ratio = sch_config.get("ckpt_ratio", 0.0)
+    if ckpt_ratio > 0.0:
+        prefix = sch_config.get("prefix", "")
+        checkpoint_method = sch_config.get("checkpoint_method", "uniform")
+        logger.info("Checkpoint ratio: %.2f", ckpt_ratio, ranks=0)
+        n_ckpt = checkpoint(
+            sch[prefix],
+            model_config,
+            ckpt_ratio=ckpt_ratio,
+            checkpoint_method=checkpoint_method,
+        )
+        logger.info("Checkpointed %d layers", n_ckpt, ranks=0)
+
+    # Pipeline parallelism.
+    if sch_config.get("pipeline_cuts", None):
+        logger.info("Generate pipeline schedule", ranks=0)
+        generate_pipeline_schedule(sch, sch_config)
+
+    return sch
+
+
+def shard_parameters(sch, model_config, sch_config):
+    model_name = model_config._name_or_path
+    logger = get_logger(model_name)
+    prefix = sch_config.get("prefix", "")
+    delay_init = sch_config.get("delay_init", True)
+    sequence_parallel = sch_config.get("sequence_parallel", False)
+    # Replace self attention with flash attention.
+    attn_op_name = sch_config.get("attn_op_name", "cuda")
+    if attn_op_name == "native_xformers":
+        logger.info("Disabled Flash Attention", ranks=0)
+    cnt, applied_attn_op_name = replace_attention(
+        sch[prefix],
+        model_config,
+        delay_init=delay_init,
+        attn_op_name=attn_op_name,
+    )
+    logger.info(
+        "Replace %d attention layers with %s op", cnt, applied_attn_op_name, ranks=0
+    )
+
+    # Replace MLP with fused kernels.
+    cnt = replace_mlp(sch[prefix], model_config, delay_init=delay_init)
+    logger.info("Replaced %d MLP layers", cnt, ranks=0)
```

Comment:
This logic is not about sharding and should not be in "shard_parameters". So do other models.